net-utils/src/token_bucket.rs

```diff
@@ -0,0 +1,384 @@
+        }
+    }
+
+    /// Fetches amount of tokens available for key.
+    ///
+    /// Returns None if no bucket exists for the key provided
+    #[inline]
+    pub fn current_tokens(&self, key: impl Borrow<K>) -> Option<u64> {
+        let bucket = self.data.get(key.borrow())?;
+        Some(bucket.current_tokens())
+    }
+
+    /// Consumes request_size tokens from a bucket at given key.
+    ///
+    /// On success, returns Ok(amount of tokens left in the bucket)
+    /// On failure, returns Err(amount of tokens missing to fill request)
+    /// If no bucket exists at key, a new bucket will be allocated, and normal policy will be applied to it
+    /// Outdated buckets may be evicted on an LRU basis.
+    pub fn consume_tokens(&self, key: K, request_size: u64) -> Result<u64, u64> {
+        let (entry_added, res) = {
+            let bucket = self.data.entry(key);
+            match bucket {
+                Entry::Occupied(entry) => (false, entry.get().consume_tokens(request_size)),
+                Entry::Vacant(entry) => {
+                    // if the key is not in the LRU, we need to allocate a new bucket
+                    let bucket = self.prototype_bucket.clone();
+                    let res = bucket.consume_tokens(request_size);
+                    entry.insert(bucket);
+                    (true, res)
+                }
+            }
+        };
+
+        if entry_added {
+            if let Ok(count) =
+                self.countdown_to_shrink
+                    .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |v| {
+                        if v == 0 {
+                            // reset the countup to starting position
+                            // thus preventing other threads from racing for locks
+                            None
+                        } else {
+                            Some(v.saturating_sub(1))
+                        }
+                    })
+            {
+                if count == 1 {
+                    // the last "previous" value we will see before counter reaches zero
+                    self.maybe_shrink();
+                    self.countdown_to_shrink
+                        .store(self.shrink_interval, Ordering::Relaxed);
+                }
+            } else {
+                self.approx_len.fetch_add(1, Ordering::Relaxed);
+            }
+        }
+        res
+    }
+
+    /// Returns approximate amount of entries in the datastructure.
+    /// Should be within ~10% of the true amount.
+    #[inline]
+    pub fn len_approx(&self) -> usize {
+        self.approx_len.load(Ordering::Relaxed)
+    }
+
+    // apply lazy-LRU eviction policy to each DashMap shard.
+    // Allowing side-effects here since overflows here are not
+    // actually possible
+    #[allow(clippy::arithmetic_side_effects)]
+    fn maybe_shrink(&self) {
+        let mut actual_len = 0;
+        let target_shard_size = self.target_capacity / self.data.shards().len();
```

Comment:
What if self.target_capacity < shards?

Reply:
We should document it and assert probably.

Reply:
This is a good catch thank you! target_size becomes zero and data structure wipes all records every time :( Will patch.